go-controller/pkg/ovn/libovsdbops/acl.go

```diff
@@ -0,0 +1,197 @@
+func findACL(nbClient libovsdbclient.Client, acl *nbdb.ACL) error {
+	if acl.UUID != "" && !IsNamedUUID(acl.UUID) {
+		return nil
+	}
+
+	acls := []nbdb.ACL{}
+	err := nbClient.WhereCache(func(item *nbdb.ACL) bool {
+		return isEquivalentACL(item, acl)
+	}).List(&acls)
+
+	if err != nil {
+		return fmt.Errorf("can't find ACL by equivalence %+v: %v", *acl, err)
+	}
+
+	if len(acls) > 1 {
+		return fmt.Errorf("unexpectedly found multiple equivalent ACLs: %+v", acls)
+	}
+
+	if len(acls) == 0 {
+		return libovsdbclient.ErrNotFound
+	}
+
+	acl.UUID = acls[0].UUID
+	return nil
+}
+
+func ensureACLUUID(acl *nbdb.ACL) {
+	if acl.UUID == "" {
+		acl.UUID = BuildNamedUUID()
+	}
+}
+
+func BuildACL(name string, direction nbdb.ACLDirection, priority int, match string, action nbdb.ACLAction, meter string, severity nbdb.ACLSeverity, log bool, externalIds map[string]string) *nbdb.ACL {
+	name = fmt.Sprintf("%.63s", name)
+	return &nbdb.ACL{
+		Name:        &name,
+		Direction:   direction,
+		Match:       match,
+		Action:      action,
+		Priority:    priority,
+		Severity:    &severity,
+		Log:         log,
+		Meter:       &meter,
+		ExternalIDs: externalIds,
+	}
+}
+
+func createOrUpdateACLOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, acl *nbdb.ACL) ([]libovsdb.Operation, error) {
+	if ops == nil {
+		ops = []libovsdb.Operation{}
+	}
+
+	err := findACL(nbClient, acl)
+	if err != nil && err != libovsdbclient.ErrNotFound {
+		return nil, err
+	}
+
+	// If ACL does not exist, create it
+	if err == libovsdbclient.ErrNotFound {
+		ensureACLUUID(acl)
+		op, err := nbClient.Create(acl)
+		if err != nil {
+			return nil, err
+		}
+		ops = append(ops, op...)
+		return ops, nil
+	}
+
+	uuid := acl.UUID
+	acl.UUID = ""
+	op, err := nbClient.Where(&nbdb.ACL{UUID: uuid}).Update(acl)
+	if err != nil {
+		return nil, err
+	}
+	ops = append(ops, op...)
+	acl.UUID = uuid
+
+	return ops, nil
+}
+
+func CreateOrUpdateACLsOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, acls ...*nbdb.ACL) ([]libovsdb.Operation, error) {
+	if ops == nil {
+		ops = []libovsdb.Operation{}
+	}
+	for _, acl := range acls {
+		var err error
+		ops, err = createOrUpdateACLOps(nbClient, ops, acl)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return ops, nil
+}
+
+func CreateOrUpdateACLs(nbClient libovsdbclient.Client, acls ...*nbdb.ACL) error {
+	ops, err := CreateOrUpdateACLsOps(nbClient, nil, acls...)
+	if err != nil {
+		return err
+	}
+
+	_, err = TransactAndCheck(nbClient, ops)
+	return err
+}
+
+func UpdateACLLoggingOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, acl *nbdb.ACL) ([]libovsdb.Operation, error) {
+	if ops == nil {
+		ops = []libovsdb.Operation{}
+	}
+
+	err := findACL(nbClient, acl)
+	if err != nil {
+		return nil, err
+	}
+
+	uuid := acl.UUID
+	acl.UUID = ""
+	op, err := nbClient.Where(&nbdb.ACL{UUID: uuid}).Update(acl, &acl.Severity, &acl.Log)
+	if err != nil {
+		return nil, err
+	}
+	ops = append(ops, op...)
+	acl.UUID = uuid
```

Comment:
Why do we store the acl UUID in a temporary variable and unset and then reset? Is this a nuance or the libovsdb `Update` method?

Reply:
Yeah, that is a bit annoying. libovsdb gets upset if we try to set a non-mutable field, like the uuid and would fail the update.

Reply:
Yes, if we set UUID for an Update it will fail. Will be fixed with https://github.com/ovn-org/libovsdb/pull/210